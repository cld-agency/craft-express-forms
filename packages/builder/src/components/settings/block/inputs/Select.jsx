import PropTypes from 'prop-types';
import React from 'react';
import translate from '../../../../functions/translator';
import BaseInput from './BaseInput';

class Select extends BaseInput {
  static propTypes = {
    ...BaseInput.propTypes,
    emptyOption: PropTypes.string,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.node,
        label: PropTypes.node,
      })
    ),
  };

  renderInput() {
    const { name, value, emptyOption, saveHandler, options = [], disabled = false } = this.props;

    return (
      <div className="select fullwidth">
        <select name={name} value={value} onChange={saveHandler} disabled={disabled}>
          {emptyOption && <option value="">{translate(emptyOption)}</option>}
          {options.map((item) => (
            <option value={item.value} key={item.value}>
              {translate(item.label)}
            </option>
          ))}
        </select>
      </div>
    );
  }
}

export default Select;
