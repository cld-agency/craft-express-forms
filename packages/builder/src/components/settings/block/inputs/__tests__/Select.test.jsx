import { shallow } from 'enzyme';
import React from 'react';
import Select from '../Select';

describe('<Select />', () => {
  let wrapper;
  let onChange;
  beforeEach(() => {
    onChange = jest.fn();
    wrapper = shallow(
      <Select
        label="A Test Select Box"
        name="test_select"
        value="two"
        saveHandler={onChange}
        options={[
          { value: 'one', label: 'Value One' },
          { value: 'two', label: 'Value Two' },
          { value: 'three', label: 'Value Three' },
        ]}
      />
    );
  });

  it('renders & displays correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('contains a textarea field', () => {
    expect(wrapper.find("select[name='test_select']")).toHaveLength(1);
  });

  it('changes value', () => {
    const value = 'three';
    const select = wrapper.find("select[name='test_select']");
    expect(select.prop('value')).toEqual('two');

    select.simulate('change', { target: { value } });
    expect(onChange).toBeCalledWith({ target: { value: 'three' } });
  });

  it('has an empty option', () => {
    const localWrapper = shallow(
      <Select
        label="A Test Select Box"
        name="test_select"
        value="two"
        saveHandler={onChange}
        emptyOption={'This has an empty option'}
        options={[
          { value: 'one', label: 'Value One' },
          { value: 'two', label: 'Value Two' },
          { value: 'three', label: 'Value Three' },
        ]}
      />
    );

    const select = localWrapper.find("select[name='test_select']");

    expect(select.find("option[value='']")).toHaveLength(1);
  });

  it('does not have an empty option', () => {
    const localWrapper = shallow(
      <Select
        label="A Test Select Box"
        name="test_select"
        value="two"
        saveHandler={onChange}
        options={[
          { value: 'one', label: 'Value One' },
          { value: 'two', label: 'Value Two' },
          { value: 'three', label: 'Value Three' },
        ]}
      />
    );

    const select = localWrapper.find("select[name='test_select']");

    expect(select.find("option[value='']")).toHaveLength(0);
  });
});
