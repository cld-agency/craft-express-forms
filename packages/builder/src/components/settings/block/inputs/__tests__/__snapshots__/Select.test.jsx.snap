// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders & displays correctly 1`] = `
<div
  className="block-row"
>
  <div
    className="label"
  >
    <WithStyles(ForwardRef(Tooltip))
      placement="bottom"
      title=""
    >
      <label
        className=""
      >
        A Test Select Box
      </label>
    </WithStyles(ForwardRef(Tooltip))>
  </div>
  <div
    className="input ltr"
  >
    <div
      className="select fullwidth"
    >
      <select
        disabled={false}
        name="test_select"
        onChange={[MockFunction]}
        value="two"
      >
        <option
          key="one"
          value="one"
        >
          Value One
        </option>
        <option
          key="two"
          value="two"
        >
          Value Two
        </option>
        <option
          key="three"
          value="three"
        >
          Value Three
        </option>
      </select>
    </div>
  </div>
</div>
`;
